### PR TITLE
Cpu tgi

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: server/text_generation_server/models/cache_manager.py [hunk 1]
```diff
@@ -25,26 +25,43 @@ def __init__(
         self.repeat_slots = repeat_slots
 
         element_size = torch.tensor([], dtype=dtype).element_size()
-        if IS_XPU_SYSTEM:
-            x = 1
+        if device == torch.device("cpu"):
```
Comment (Narsil) [inline server/text_generation_server/models/cache_manager.py:28]: Why ?

This breaks a lot of assumptions for other places in code.
We are experimenting with switching to flash decoding which has this layout, but it doesn't feel right to change the layout based on device.

File: server/text_generation_server/models/flash_llama.py [hunk 2]
```diff
@@ -35,9 +35,13 @@ def __init__(
...
             dtype = torch.float16 if dtype is None else dtype
-        elif IS_XPU_SYSTEM:
-            device = torch.device(f"xpu:{rank}")
-            dtype = torch.float16 if dtype is None else dtype
+        elif IS_IPEX_AVAILABLE:
+            if IS_XPU_SYSTEM:
+                device = torch.device(f"xpu:{rank}")
+                dtype = torch.float16 if dtype is None else dtype
+            else:
+                device = torch.device("cpu")
```
Comment (Narsil) [inline server/text_generation_server/models/flash_llama.py:43]: IS flash attention + paged implemented for IPEX + CPU ? 
Or any CPU ?

File: server/text_generation_server/utils/layers.py [hunk 3]
```diff
@@ -875,19 +878,16 @@ def load(cls, prefix, weights, eps=1e-6):
...
-                residual_out = hidden_states
+            if IS_IPEX_AVAILABLE:
                 out = ipex.llm.functional.add_rms_norm(
                     residual,
                     hidden_states,
                     self.weight,
                     None,
                     self.variance_epsilon,
-                    True,
+                    residual is not None,
```
Comment (Narsil) [inline server/text_generation_server/utils/layers.py:888]: What is this `True` flag ?

Is the call calling into FFI boundary (I'm wondering why the callsite cannot determine the potential None value of `residual` directly without the need for this extra flag)

**Review by Narsil — COMMENTED** [review]
Nice !

File: server/text_generation_server/models/flash_llama.py (hunk 2 above)
Comment (sywangyi) [reply server/text_generation_server/models/flash_llama.py:43]: for IPEX+CPU, IPEX provide same pageattention+flash attention op for xpu+cpu

File: server/text_generation_server/models/cache_manager.py (hunk 1 above)
Comment (sywangyi) [reply server/text_generation_server/models/cache_manager.py:28]: cpu and xpu in ipex implement difference layout for pageattention.

File: server/text_generation_server/utils/layers.py (hunk 3 above)
Comment (sywangyi) [reply server/text_generation_server/utils/layers.py:888]: user could control if residual will be updated in place by this parameter.

**Comment by sywangyi** [comment]
@yao-matrix @jianan-gu
Hi, @Narsil  I rebase the PR to latest main, please help review

**Comment by sywangyi** [comment]
Docker image build for xpu:
sudo docker build -t tgi_xpu --target xpu  . -f Dockerfile_intel
Docker image build for cpu:
sudo docker build -t tgi_cpu --target cpu  . -f Dockerfile_intel

text_generation_launcher for xpu:
sudo docker run --rm --shm-size 1g -p 8080:80  -e HUGGING_FACE_HUB_TOKEN=xxxxx --cap-add=sys_nice --ipc=host -v ~/data:/data  --device /dev/dri --privileged  tgi_xpu --model-id=meta-llama/Llama-2-7b-chat-hf --cuda-graphs 0

text_generation_launcher for cpu:
sudo docker run --rm --shm-size 1g -p 8080:80  -e HUGGING_FACE_HUB_TOKEN=xxxxx --cap-add=sys_nice --ipc=host -v ~/data:/data  tgi_cpu --model-id=meta-llama/Llama-2-7b-chat-hf --cuda-graphs 0

**Comment by sywangyi** [comment]
@Narsil  rebase cpu support in latest main. please help review, CPU and XPU share same ipex.llm.xxx interface. only difference is cache layout.

**Comment by sywangyi** [comment]
may need update https://github.com/huggingface/text-generation-inference/blob/main/.github/workflows/build.yaml to add intel-cpu docker image build. build command is like sudo docker build -t tgi_cpu --build-arg PLATFORM=cpu . -f Dockerfile_intel

**Comment by yao-matrix** [comment]
@Narsil , since vLLM optimized cpu backend already merged, could you help review the TGI counterpart, and leave any comments , and we will resolve them ASAP.

File: Dockerfile_intel [hunk 4]
```diff
@@ -36,7 +38,8 @@ RUN cargo build --profile release-opt
 
 
 # Text Generation Inference base image for Intel
-FROM intel/intel-extension-for-pytorch:2.1.30-xpu as base
+
+FROM intel/intel-extension-for-pytorch:2.1.30-xpu as xpu
```
Comment (mfuntowicz) [inline Dockerfile_intel:42]: Should we update PyTorch to a more recent version? Or do you prefer to stick to this one?

**Review by mfuntowicz — DISMISSED** [review]
LGTM - What are the requirements regarding testing? What kind of systems are you targeting mainly? SPR?

File: Dockerfile_intel (hunk 4 above)
Comment (sywangyi) [reply Dockerfile_intel:42]: for xpu. we stick to this one currently, We plan to release new version in end of July. and I will update the dockerfile at that time.

**Comment by sywangyi** [comment]
> LGTM - What are the requirements regarding testing? What kind of systems are you targeting mainly? SPR?

Yes, SPR. For earlier CPUs, it can run but not best performance since no AMX.

**Comment by sywangyi** [comment]
@Narsil could you revisit the PR, since @mfuntowicz  has approved it.

**Comment by yao-matrix** [comment]
@Narsil , pls help review, we hope it can merge ASAP, then we can add more features like AWQ support based on it. Thx.

File: server/text_generation_server/models/flash_causal_lm.py [hunk 5]
```diff
@@ -773,21 +773,38 @@ def init_kv_cache(
...
-                    device=device,
-                ),
-            )
-            for _ in range(num_layers)
-        ]
+        if IPEX_AVAIL and SYSTEM == "cpu":
+            self.kv_cache = [
+                (
+                    torch.empty(
+                        (num_blocks, num_heads, BLOCK_SIZE, head_size),
```
Comment (Narsil) [inline server/text_generation_server/models/flash_causal_lm.py:780]: Why the difference with XPU ? Aren't the the same kernel ?

**Review by Narsil — COMMENTED** [review]
In order to speed things up I'll merge this and fix the nits in follow up PRs.

**Review by Narsil — APPROVED** [review]

**Pull request merged**